docs/source/configuration/validation.md

```diff
@@ -0,0 +1,297 @@
+
+```python
+from plone.supermodel import model
+from zope import schema
+
+class IMyContent(model.Schema):
+    directives.widget(
+        "customField",
+        frontendOptions={
+            "format": "url",
+        },
+    )
+    customField = schema.TextLine(
+        title="Custom URL field",
+        required=False,
+    )
+    # Rest of your content type definition
+```
+
+The response from `plone.restapi` will be something like the following.
+It is slightly different from blocks JSON schema, but the validation engine will behave the same.
+The `urlValidator` method validator will be applied for the content type field `customField` from the earlier example.
+
+```json
+{
+  "properties": {
+    "customField": {
+      "title": "Custom URL field",
+      "widgetOptions": {
+        "frontendOptions": {
+          "format": "url"
+        }
+      }
+    }
+  }
+}
+```
+
+
+### Advanced scenarios
+
+In case you need more granularity, and you don't have access to modify the existing implementation of the JSON schema definitions for existing content types, blocks, or forms, you can use the following advanced validator registrations, using `field`, `widget`, `behaviorName`, or `blockType` validator registrations.
+
+
+#### Field `type` validators
+
+Field `type` validators are applied depending on the specified `type` of the field in the JSON schema from content types, forms, or blocks.
+
+You should specify the `type` in the JSON schema of the block.
+In a content type, it is included in the default serialization of the field.
+
+If a field does not specify `type`, it assumes a `string` type as validator.
+
+The next example shows how to define the JSON schema in a block.
+
+```ts
+let blockSchema = {
+  // ... fieldset definition in here
+  properties: {
+    ...schema.properties,
+    customField: {
+      title: 'My custom field',
+      description: '',
+      type: 'integer',
+      maximum: 30
+    },
+  },
+  required: [],
+};
+```
+
+```ts
+config.registerUtility({
+  type: 'validator',
+  name: 'maximum',
+  dependencies: {
+    fieldType: 'integer',
+  },
+  method: maximumValidator,
+})
+```
+
+
+#### Field `widget` validators
+
+Field `widget` validators are applied depending on the specified `widget` of the field.
+You should specify the `widget` either in the JSON schema of the block or as additional data in the content type definition.
+
+The following example shows how to specify the `widget` either in the JSON schema of the block.
+
+```ts
+let blockSchema = {
+  // ... fieldset definition in here
+  properties: {
+    ...schema.properties,
+    customField: {
+      title: 'My custom field',
+      description: '',
+      widget: 'phoneNumber',
+    },
+  },
+  required: [],
+};
+```
+
+```ts
+config.registerUtility({
+  type: 'validator',
+  name: 'phoneNumber',
+  dependencies: {
+    widgetName: 'phoneNumber',
```

Comment:
Why is it `widget` in the schema but `widgetName` here?

Reply:
Moved to use `widget` only.